### PR TITLE
[SPARK-50405][SQL] Handle collation type coercion of complex data types properly

### PR DESCRIPTION
**Pull request by stefankandic**
### What changes were proposed in this pull request?
This pull request generalizes collation type coercion to support not just casting all children to a single string type, but also handling complex data types such as structs, maps, and arrays (arrays partially worked already).

The core idea is to recursively analyze the entire data type of an expression, annotating each StringType within it with the highest-priority collation and its strength. This annotation propagates upward through the expression tree. Once the root of the expression is reached, the annotations are removed, and the expression is cast to the desired data type.

For the root expression `e`, the collation data type context is computed by first calculating the context for all its children and then merging those results into the data type of `e`.

### Why are the changes needed?
In #48663, a new approach to calculating collation precedence was introduced. This approach recursively examines the children of an expression and propagates the collation with the highest priority upward.

However, the current implementation of collation coercion is limited to determining the StringType that all children should be cast to. This approach falls short when dealing with complex types like structs, maps, and arrays, which can also contain collations. To address this limitation, we need a more general mechanism that allows coercion of any data type, not just simple strings.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
With new unit tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by stefankandic** [comment]
@cloud-fan please take a look when you get the chance.

You raised some valid concerns in the last PR about structs and maps so I think I was able to figure out a nice way to have it work for all data types :)

**Review by MaxGekk — COMMENTED** [review]
@stevomitric @uros-db @ivanjevtic-db @vladanvasi-db Please, review this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:1]: quite a few changes here, do we have benchmarks to detect possible "regressions" (i.e. any slowdown as opposed to the previous approach)?

File: sql/core/src/test/scala/org/apache/spark/sql/collation/CollationTypePrecedenceSuite.scala [hunk 1]
```diff
@@ -20,6 +20,7 @@ package org.apache.spark.sql.collation
 import org.apache.spark.SparkThrowable
 import org.apache.spark.sql.{DataFrame, QueryTest, Row}
 import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.{ArrayType, DataType, IntegerType, MapType, StringType, StructField, StructType}
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/collation/CollationTypePrecedenceSuite.scala:23]:
```suggestion
import org.apache.spark.sql.types._
```

**Review by stevomitric — APPROVED** [review]

**Review by vladanvasi-db — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala
Comment (stefankandic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:1]: Cases covered previously should remain unaffected. Plain strings will continue to be coerced exactly as before. However, increasing the nesting of complex types will naturally introduce additional calculations, which is unavoidable in such scenarios.

**Review by dejankrak-db — APPROVED** [review]

**Review by uros-db — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala [hunk 2]
```diff
@@ -450,7 +450,10 @@ class CollationSQLRegexpSuite
         },
         condition = "DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE",
         parameters = Map(
-          "sqlExpr" -> "\"regexp_replace(collate(ABCDE, UNICODE_CI), .c., FFF, 1)\"",
+          "sqlExpr" ->
+            """
+              |"regexp_replace(collate(ABCDE, UNICODE_CI), .c., 'FFF' collate UNICODE_CI, 1)"
+              |""".stripMargin.trim,
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:456]:
```suggestion
          "sqlExpr" ->
            """"regexp_replace(collate(ABCDE, UNICODE_CI), .c., 'FFF' collate UNICODE_CI, 1)"""",
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala [hunk 3]
```diff
@@ -164,31 +160,89 @@ object CollationTypeCoercion {
...
+
+      // handle the base cases first
+      case _ if baseCase.isDefinedAt((leftType, rightType)) =>
+        Option(baseCase(leftType, rightType))
+
+      case _ if leftType == rightType =>
+        Some(leftType)
+
+      case (ArrayType(leftElemType, nullable), ArrayType(rightElemType, _)) =>
+        mergeStructurally(leftElemType, rightElemType)( baseCase).map(ArrayType(_, nullable))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:218]:
```suggestion
        mergeStructurally(leftElemType, rightElemType)(baseCase).map(ArrayType(_, nullable))
```

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:218]: Could you clarify merging `nullable`, so, if the left array type has `nullable` is `false`, but the right one has `true`, you propose to set it to `false`? I would expect something like `leftNullable || rightNullable`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala [hunk 4]
```diff
@@ -164,31 +160,89 @@ object CollationTypeCoercion {
...
+
+      case (MapType(leftKey, leftValue, nullable), MapType(rightKey, rightValue, _)) =>
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:220]: the same question regarding `nullable`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala [hunk 5]
```diff
@@ -223,151 +277,195 @@ object CollationTypeCoercion {
     val collationContextWinner = expressions.foldLeft(findCollationContext(expressions.head)) {
       case (Some(left), right) =>
         findCollationContext(right).flatMap { ctx =>
-          collationPrecedenceWinner(left, ctx)
+          mergeWinner(left, ctx)
         }
       case (None, _) => return None
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:282]: Not related to the PR, but `return` is not necessary, it seems.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala [hunk 6]
```diff
@@ -223,151 +277,195 @@ object CollationTypeCoercion {
...
+    collationContextWinner
   }
 
   /**
    * Tries to find the collation context for the given expression.
    * If found, it will also set the [[COLLATION_CONTEXT_TAG]] on the expression,
    * so that the context can be reused later.
    */
-  private def findCollationContext(expr: Expression): Option[CollationContext] = {
+  private def findCollationContext(expr: Expression): Option[DataType] = {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:292]: Should we adjust the comment above regarding the return type?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala (hunk 3 above)
Comment (stefankandic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:218]: Let's say you have an expression `e` with children `c1, c2, c3`. You will find the least common type for them `t = LCT(c1, c2, c3)`. Then you should apply that type to all children (bear in mind that some of these children can be complex while others can be primitive types).

When applying the new type `t` we should not change the nullability of the type that the children had originally, and that's why I am doing it this way. Not sure if I could have named the parameters better, but the `left` and `right` type are not equal, we are trying to merge the collations of the right one into the left and keep everything else the same.

It wouldn't make sense for one of the children to suddenly become nullable after coercion or the other way around.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala (hunk 5 above)
Comment (stefankandic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala]: nice catch, removed the return

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala (hunk 6 above)
Comment (stefankandic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:292]: done!

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @stefankandic and @uros-db @stevomitric @dejankrak-db @vladanvasi-db for review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala (hunk 3 above)
Comment (tedyu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala]: Makes sense.
Maybe name the parameters `original` and `update`: Implies right contains updates applied to left.

**Pull request closed without merging**